packages/dds/tree/src/feature-libraries/flex-map-tree/mapTreeNode.ts

```diff
@@ -0,0 +1,744 @@
+
+// #region Nodes
+
+/** A node's parent field and its index in that field */
+interface LocationInField {
+	readonly parent: MapTreeField<FlexAllowedTypes>;
+	readonly index: number;
+}
+
+/**
+ * A readonly implementation of {@link FlexTreeNode} which wraps a {@link MapTree}.
+ * @remarks Any methods that would mutate the node will fail,
+ * as will the querying of data specific to the {@link LazyTreeNode} implementation (e.g. {@link FlexTreeNode.context}).
+ * MapTreeNodes are unconditionally cached -
+ * when retrieved via {@link getOrCreateMapTreeNode}, the same {@link MapTree} object will always produce the same `MapTreeNode` object.
+ */
+export class MapTreeNode<TSchema extends FlexTreeNodeSchema> implements FlexTreeNode {
+	public readonly [flexTreeMarker] = FlexTreeEntityKind.Node as const;
+
+	private readonly events = createEmitter<FlexTreeNodeEvents>();
+
+	/**
+	 * Create a new MapTreeNode.
+	 * @param schema - the schema of this node
+	 * @param mapTree - the {@link MapTree} containing the data that this node wraps
+	 * @param location - the parentage of this node, if it is being created underneath an existing node and field, or undefined if not
+	 * @remarks This class (and its subclasses) should not be directly constructed outside of this module.
+	 * Instead, use {@link getOrCreateMapTreeNode} to create a MapTreeNode from a {@link MapTree}.
+	 */
+	public constructor(
+		public readonly schema: TSchema,
+		public readonly mapTree: MapTree,
+		private location: LocationInField | undefined,
+	) {
+		assert(!nodeCache.has(mapTree), "A node already exists for the given MapTree");
+		nodeCache.set(mapTree, this);
+
+		// Fully demand the tree to ensure that parent pointers are present and accurate on all nodes.
+		// When a MapTreeNode is constructed, its MapTree may contain nodes (anywhere below) that map (via the `nodeCache`) to pre-existing MapTreeNodes.
+		// Put another way, for a given MapTree, some ancestor MapTreeNode can be created after any number of its descendant MapTreeNodes already exist.
+		// In such a case, the spine of nodes between the descendant and ancestor need to exist in order for the ancestor to be able to walk upwards via the `parentField` property.
+		// This needs to happen for all MapTreeNodes that are descendants of the ancestor MapTreeNode.
+		// Demanding the entire tree is overkill to solve this problem since not all descendant MapTree nodes will have corresponding MapTreeNodes.
+		// However, demanding the full tree also lets us eagerly validate that there are no duplicate MapTrees (i.e. same MapTree object) anywhere in the tree.
+		this.walkTree();
+	}
+
+	public get type(): TreeNodeSchemaIdentifier {
+		return this.schema.name;
+	}
+
+	/**
+	 * Set this node's parentage (see {@link FlexTreeNode.parentField}).
+	 * @remarks A node may only be adopted to a new parent one time, though this method may be called idempotently with the same parent and index.
```

Comment:
For Josh's understanding: what is the benefit of allowing this to be called multiple times with the same params? Vs just unconditionally failing if this is called multiple times.

Reply:
Oh yea... it used to be that I didn't cache the MapTreeFields. So, if you constructed a field object multiple times for the same children, it was nice to be able to call "adopt" unconditionally on each one. But now that's not needed, so I'll change the assert to be simpler and just say "don't adopt twice".